### ISSUE TITLE
Create run-inspect-viz skill for visualizing experiment results

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/cruijff_kit#170]
# Create `run-inspect-viz` skill for visualizing experiment results

## Context

The `analyze-experiment` skill (currently planned) needs a worker skill to generate visualizations from Inspect AI evaluation results. The `run-inspect-viz` skill will use [inspect-viz](https://github.com/meridianlabs-ai/inspect_viz) to create interactive HTML visualizations that help researchers understand and compare experimental results.

This follows the established architecture pattern where:
- `analyze-experiment` orchestrates analysis workflows
- `run-inspect-viz` handles visualization generation (worker skill)
- Can also be run standalone for quick visualization needs

### Input
- **Experiment directory** containing:
  - `experiment_summary.md` (for experiment design context)
  - Run directories with `eval/logs/*.eval` files
  - Status tracking showing completed evaluations

### Output
- **Interactive HTML visualizations** including:
  1. **Custom plots based on experiment design**:
     - Automatically infer visualization type from experimental variables
     - Examples:
       - LoRA rank comparison → bar plot or line plot
       - Learning rate sweep → line plot with multiple series
       - Model size comparison → grouped bar chart
       - Multi-factor experiments → heatmaps or faceted plots

  2. **Appendix with all pre-built views** from inspect-viz:
     - `scores_by_model()` - Compare scores across models
     - `scores_by_task()` - Performance across tasks
     - `scores_timeline()` - Score progression over time
     - `scores_heatmap()` - Score matrix visualization
     - `scores_by_factor()` - Boolean factor comparisons
     - `sample_tool_calls()` - Tool usage patterns (if applicable)
     - Other relevant views from `inspect_viz.view.beta`

### Workflow

1. **Locate experiment** - Find experiment directory (current dir or user-specified)
2. **Read experiment_summary.md** - Extract:
   - Experimental design (variables, factors, levels)
   - Run configurations
   - Evaluation tasks
   - Research questions
3. **Discover .eval files** - Scan `{run_dir}/eval/logs/*.eval` for all evaluations
4. **Load evaluation data** - Use Inspect AI's dataframe API to load results
5. **Infer visualization strategy** - Based on:
   - Which parameters vary across runs
   - Number of tasks evaluated
   - Metrics available in results
   - Experiment type (factorial, sweep, ablation, etc.)
6. **Generate custom plots** - Create experiment-specific visualizations
7. **Generate pre-built views** - Apply all relevant inspect-viz views
8. **Write HTML output** - Save interactive visualizations to experiment directory
9. **Create log** - Document visualization process in `run-inspect-viz.log`
10. **Report summary** - Show user what was generated and where

### Output Structure

**Approach:** Generate separate HTML files (one per visualization) with a navigation index, rather than a single-page dashboard. This keeps files lightweight and loads faster.

```
{experiment_dir}/
├── experiment_summary.md
├── run-inspect-viz.log
├── visualizations/
│   ├── index.html              # Navigation with links to all visualizations
│   ├── experiment_design.html  # Custom plots for this specific experiment
│   ├── scores_by_model.html    # Pre-built view
│   ├── scores_by_task.html     # Pre-built view
│   ├── scores_timeline.html    # Pre-built view
│   ├── scores_heatmap.html     # Pre-built view
│   └── ...                     # Other relevant pre-built views
└── rank*/
    └── eval/logs/*.eval
```

**Future:** Could optionally generate a Quarto Dashboard for single-page experience.

### Dependencies

**Required:**
- `inspect-viz` - Visualization library
  ```bash
  pip install inspect-viz
  ```
- `inspect-ai` - Already a cruijff_kit dependency (provides eval log reading)

**Optional:**
- `playwright` - For PNG export (if implementing static image output in future)
  ```bash
  pip install playwright
  playwright install
  ```

**Note:** inspect-viz automatically installs its dependencies including:
- `pandas` >= 2.2.2 (dataframe handling)
- `pyarrow` >= 15.0.0 (parquet file reading)
- `anywidget` >= 0.9.0 (interactive widgets)
- `narwhals` >= 1.15.1 (dataframe compatibility layer)

### Data Loading Pattern

Use Inspect AI's dataframe functions to load evaluation results:

**Primary approach (for experiment comparisons):**
```python
from inspect_ai.log import evals_df, prepare, model_info, task_info
from inspect_viz import Data
import glob

# Locate all evaluation logs
eval_logs = glob.glob(f"{experiment_dir}/*/eval/logs/*.eval")

# Load evaluation-level data (one row per eval run)
evals = evals_df(eval_logs)
evals = prepare(evals, model_info(), task_info())  # Add metadata

# Create inspect-viz Data object
evals_data = Data.from_dataframe(evals)
```

**Alternative granularities (for specific analyses):**
- `samples_df()` - Sample-level data for error analysis and score distributions
- `messages_df()` - Message-level for conversation pattern analysis
- `events_df()` - Event-level for tool call visualization

**Reference:** https://inspect.aisi.org.uk/dataframe.html

### Visualization Inference Logic

**Approach:** Parse `experiment_summary.md` to identify which parameters vary across runs, then generate plots appropriate for the experimental design.

**Visualization strategy based on experimental variables:**

**Single varying parameter:**
- Bar chart sorted by score
- Example: LoRA rank (8, 16, 32, 64) → horizontal bar chart comparing configurations

**Two varying parameters:**
- Heatmap showing all combinations
- Grouped bar chart or line plot with multiple series
- Example: LoRA rank × learning rate → 2D heatmap with scores as cell colors

**Multiple epochs:**
- Line plot showing training progression
- Example: Evaluations at epoch 0, 1, 2 → learning curves for each configuration

**Multiple tasks:**
- Separate plots per task or unified comparison view
- Example: 3 different evaluation tasks → 3 plots or faceted layout

### HTML Generation

Use `write_html()` to save visualizations:

```python
from inspect_viz.plot import write_html
from inspect_viz.view.beta import scores_by_model

# Generate and save pre-built views
viz = scores_by_model(evals_data)
write_html("visualizations/scores_by_model.html", viz)

# Generate and save custom plots (based on experimental design)
custom_viz = create_experiment_plot(evals_data, experiment_design)
write_html("visualizations/experiment_design.html", custom_viz)
```

**Key functions:**
- `write_html(path, component)` - Save any plot, view, or layout as standalone HTML (preserves interactivity)
- `write_png(path, component)` - Save as static PNG (requires playwright, for future use)

**Navigation index:**

Create a simple `index.html` with:
- Experiment metadata (name, date, run count)
- Links to custom visualizations
- Links to pre-built views
- Clean styling for easy navigation

**Reference:** https://meridianlabs-ai.github.io/inspect_viz/publishing-plots.html

**Future enhancement:** Consider using Quarto Dashboards for single-page dashboard experience.

### Called by `analyze-experiment` (orchestrator)
- `analyze-experiment` calls `run-inspect-viz` after loading data
- Passes experiment directory path
- Incorporates visualizations into analysis report

### Standalone usage
```bash
# User runs directly for quick visualization
cd /path/to/experiment
# Invoke run-inspect-viz skill via Claude Code
```

### Relationship to existing skills
- **Input from**: `run-inspect` (produces .eval files)
- **Input from**: `design-experiment` (experiment_summary.md)
- **Called by**: `analyze-experiment` (future orchestrator)
- **Similar to**: `scaffold-inspect`, `run-inspect` (worker skills)

## Logging

Create `run-inspect-viz.log` with:

```
[YYYY-MM-DD HH:MM:SS] DISCOVER_EXPERIMENT: Found experiment
Details: /path/to/experiment
Result: Found 8 runs with evaluation results

[YYYY-MM-DD HH:MM:SS] READ_DESIGN: Parsing experiment_summary.md
Details: Experimental factors: LoRA rank (8,16,32,64), Learning rate (1e-5, 5e-5)
Result: Identified 2×4 factorial design

[YYYY-MM-DD HH:MM:SS] LOAD_EVALS: Loading evaluation logs
Details: Found 8 .eval files
Result: Loaded 480 samples across 8 runs

[YYYY-MM-DD HH:MM:SS] INFER_VISUALIZATIONS: Determining plot strategy
Details: 2 varying parameters, 1 task, 1 epoch per run
Result: Will create heatmap and grouped bar chart

[YYYY-MM-DD HH:MM:SS] GENERATE_CUSTOM: Creating experiment-specific plots
Details: Generated 2 custom visualizations
Result: Saved to visualizations/experiment_design.html

[YYYY-MM-DD HH:MM:SS] GENERATE_VIEWS: Creating pre-built views
Details: Generated 6 standard views
Result: All views saved to visualizations/

[YYYY-MM-DD HH:MM:SS] CREATE_INDEX: Building navigation
Details: Created index.html with links to all visualizations
Result: Dashboard ready at visualizations/index.html

[YYYY-MM-DD HH:MM:SS] COMPLETE: Visualization complete
Summary: Generated 8 HTML files (2 custom + 6 views)
Location: {experiment_dir}/visualizations/
```

## Error Handling

**If no .eval files found:**
- Report error
- Check if evaluations have been run
- Suggest running `run-inspect` first

**If experiment_summary.md missing:**
- Warn but continue
- Generate only pre-built views (no custom plots)
- Note limitation in summary

**If inspect-viz not installed:**
- Report clear installation instructions
- Exit gracefully

**If evaluation data malformed:**
- Log which files couldn't be loaded
- Continue with valid data
- Report issues in summary

## Success Criteria

- [ ] Loads all .eval files from experiment directory
- [ ] Parses experiment_summary.md to understand experimental design
- [ ] Infers appropriate custom visualizations based on varying parameters
- [ ] Generates all relevant pre-built inspect-viz views
- [ ] Creates interactive HTML files with navigation index
- [ ] Provides detailed logging and clear error messages
- [ ] Works standalone and as worker skill for analyze-experiment

## Example Output Summary

```markdown
## Run Inspect-Viz Complete

Successfully generated visualizations for:
`/scratch/gpfs/MSALGANIK/mjs3/cap_4L_lora_lr_sweep/`

### Custom Visualizations (2)

✓ LoRA Rank × Learning Rate Heatmap
  - Shows accuracy scores across 8 configurations
  - Interactive filtering and tooltips

✓ Score Comparison Bar Chart
  - Horizontal bars sorted by performance
  - Includes error bars from evaluation

### Pre-Built Views (6)

✓ Scores by Model
✓ Scores by Task
✓ Scores Heatmap
✓ Scores Timeline
✓ Scores by Factor
✓ Sample Heatmap

### Output Location

All visualizations saved to:
`/scratch/gpfs/MSALGANIK/mjs3/cap_4L_lora_lr_sweep/visualizations/`

**View results:**
Open `visualizations/index.html` in a browser to explore all visualizations interactively.

See `run-inspect-viz.log` for detailed generation log.
```

## Future Enhancements

- Export static PNG versions alongside HTML
- Generate Jupyter notebook with embedded visualizations
- Statistical significance testing visualization
- Cost analysis plots (GPU hours vs performance)
- Training metrics integration (from SLURM logs)
- Automated insight generation (LLM-based analysis of plots)
- Dashboard mode with all plots in single page

## References

- inspect-viz documentation: https://meridianlabs-ai.github.io/inspect_viz/
- inspect-viz GitHub: https://github.com/meridianlabs-ai/inspect_viz
- Inspect AI dataframe API: https://inspect.aisi.org.uk/dataframe.html
- Related: `analyze-experiment` skill (planned orchestrator)
- Related: `run-inspect` skill (produces input data)

## Labels

- enhancement
- skill
- visualization
- priority: medium

## Assignee

(Assign as appropriate)

## Milestone

(Add to appropriate milestone for analyze-experiment feature set)
